### ISSUE TITLE
eth_getTransactionByHash rpc call issue

### ISSUE THREAD

**Issue by martinboehm** [etclabscore/core-geth#399]
Hi there,

we are using `CoreGeth/v1.12.0-stable-c5ef3a9e/linux-amd64/go1.16.2` on Linux and have fully synchronized database. 

Unfortunately some requests to `eth_getTransactionByHash` return empty value instead of the transaction data.
For example all transactions in block `10516117`, like txid `0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa`.
The txid does exist and the transaction can fetched by using `getBlock` request. The code showing the issue, as run in the browser console:
```
var etc = new Web3('ws://ourserver');
await etc.eth.getTransaction('0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa');
> null
await etc.eth.getBlock(10516117);
> {
    "number": 10516117,
    "hash": "0x83d38d9642469a1f8e7c3533a54f14f8871372bea01d343e9e357334883c0e12",
    "parentHash": "0x77677f014d565506cc63f6e13435a6df501c82168efe1df0398df5eab24b2e75",
    "nonce": "0x5e818c6002994f9a",
    "mixHash": "0x4c67ea3f703697b25eb5a84be9c2b2d054f67a658969af155f80296487b6c206",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "stateRoot": "0xcc279e05013a54fe503179d50524f0ac3431c43dad7209cfb166481541a56fbd",
    "miner": "0xD3E2A80c4aD7656d0333205aF1c6CA01E04132E2",
    "difficulty": "99485594757006",
    "totalDifficulty": "974014593151710369792",
    "extraData": "0x706f6f6c2e65787072657373",
    "size": 868,
    "gasLimit": 8000000,
    "gasUsed": 63000,
    "timestamp": 1591169397,
    "transactionsRoot": "0xbc18f483a8b6810f255ffbba9ae9b16101a03d1a4f3d7d22bc43695116e498f2",
    "receiptsRoot": "0x251f2cb798e965c5d9b11c882f37c69fd2c42b314fabe64d2b4998c76eb93ae8",
    "transactions": [
        "0x906464e331e815568c092480ca4e05cd56705946bfbb67d907e536c81f457af0",
        "0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa",
        "0xe4846edd7b02955a6a55153aaae597bdd242829bb34893704092faa950381215"
    ],
    "uncles": []
}

```

The content of the websocket message eth_getTransactionByHash:
````
request:
{
    "jsonrpc":"2.0",
    "id":10,
    "method":"eth_getTransactionByHash",
    "params":["0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa"]
}

response:
{
    "jsonrpc":"2.0",
    "id":10,
    "result":null
}
````

Could you please give us a hint what is the issue?

**Comment by meowsbits** [comment]
Hey there @martinboehm .

I think this may be an issue with your transaction indexes. You should be able to resolve this by using `--txlookuplimit=0`. 

```
./build/bin/geth --help|grep txlook
  --txlookuplimit value               Number of recent blocks to maintain transactions index for (default = about one year, 0 = entire chain) (default: 2350000)
```

I've quoted the relevant release notes from ethereum/go-ethereum on this topic below. This change was included and released with CoreGeth beginning with the 1.12+ series.

> Geth 1.10.0 contains some changes which remove unnecessary data in the blockchain database. In particular, Geth no longer keeps transaction inclusion info for all transactions, and instead limits the storage of inclusion records to one year. For application developers, this change means that very old transactions can no longer be accessed by hash. Note: if you would like to
disable this behavior and keep inclusion information for all historical transactions, you can re-enable indexing using the --txlookuplimit=0 command-line flag. 
> https://github.com/ethereum/go-ethereum/releases/tag/v1.10.0

Using this flag, I was unable to reproduce your issue, as shown below.

```
echo '{     "jsonrpc":"2.0",     "id":10,    "method":"eth_getTransactionByHash",     "params":["0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa"] }' | websocat -n1 ws://localhost:8546
{"jsonrpc":"2.0","id":10,"result":{"blockHash":"0x83d38d9642469a1f8e7c3533a54f14f8871372bea01d343e9e357334883c0e12","blockNumber":"0xa07695","from":"0xbb8a004bc17badf11fb2cd93c20ca5b0512b148c","gas":"0x5208","gasPrice":"0x3b9aca00","hash":"0xefac823c2c440bbeda7541dcf70a244bcf9aea706331cbb199c9520183745faa","input":"0x","nonce":"0x67","to":"0x16fd8d84a3300629f0e7d275bff53360af4fee43","transactionIndex":"0x1","value":"0xde1ebd1de5b73ed","type":"0x0","v":"0x9e","r":"0x5f0aa94c6a0916be90fd933819bdaf33f915e3adcab6bc1fa15fe1e29bd8fd32","s":"0x5931a0e6782855822943728289fa81b0e362b731f28da3bac0de23ae5a0c6fa8"}}
```

**Comment by martinboehm** [comment]
@meowsbits Thank you, it fixed it!
